Mark certain symbols as .hidden in arm64 codegen, possibly fixes MPR#7585

infinity0 opened the pull request:
binutils 2.29 arm64 linker is more strict about when to perform pc-relative relocations for global symbols. Specifically, they must be marked as .hidden. ocaml arm64 codegen does not do this, so tests/lib-dynlink-native/ fails as some symbols were not relocated. This patch fixes that by marking (hopefully) the appropriate symbols as .hidden.

This patch builds and tests successfully, but should be reviewed carefully because neither me nor the original author is clear on why exactly it works. See the discussion in https://caml.inria.fr/mantis/view.php?id=7585 for details, and especially the last few comments starting from 0018143.

The change to `symbol_in_current_unit` may also justify renaming that function. (It is only used in the arm64 codegen and nowhere else.)

Commit:
Mark certain symbols as .hidden in arm64 codegen, possibly fixes MPR#7585

binutils 2.29 arm64 linker is more strict about when to perform pc-relative
relocations for global symbols. Specifically, they must be marked as .hidden.
ocaml arm64 codegen does not do this, so tests/lib-dynlink-native/ fails as
some symbols were not relocated. This patch fixes that by marking (hopefully)
the appropriate symbols as .hidden.

xavierleroy commented:
Thank you very much for this proposal.  There are certainly many possible good uses of hidden symbols in OCaml-generated code, see #1207 for another unrelated example.  In this particular case I don't yet understand the implications of declaring those symbols hidden.  Can you or anyone else explain?

The issue boils down to OCaml-ARM64's desire (actually, my desire) to refer to locally-defined-but-globally-visible symbols directly rather than indirectly through the PLT.  Maybe this desire makes no sense in a world ruled by binutils, in which case we should just give up and use PLT-indirect references for everything.

rwmjones commented:
Fedora also hit this problem on aarch64 (originally reported [here](https://caml.inria.fr/mantis/view.php?id=7585)).  I can report that this patch fixes the problem even though I am also not in a position to comment on its correctness.

I only rebuilt and tested the base compiler.  Next week [we expect to rebuild the whole OCaml chain using 4.05](https://pagure.io/releng/issue/6906) and that should provide more, or less, confidence.

mshinwell commented:
Jiong Wang's explanation on MPR#7585 sounds correct to me (many thanks for investigating this).  Globally-visible symbols in shared libraries might be subject to copy relocations; if you have a copy relocation then the symbol will resolve everywhere (including in the shared library) to the copy in the executable, which means that you have to use a GOT sequence in the shared library or you'll be looking at a stale variable if the data is mutable.

I will review the proposed patch in detail later this week.

rwmjones commented:
So we found another similar-ish problem compiling coq.  Again this only happens on aarch64, not on any other architecture:
```
rm -f theories/Init/Datatypes.glob
bin/coqtop -boot  -native-compiler -compile theories/Init/Datatypes.v  -noinit -R theories Coq
File "./theories/Init/Datatypes.v", line 13, characters 0-38:
Error: while loading
/builddir/build/BUILD/coq-8.6/plugins/syntax/nat_syntax_plugin.cmxs:
error loading shared library: /builddir/build/BUILD/coq-8.6/plugins/syntax/nat_syntax_plugin.cmxs: undefined symbol: camlLoc__3
make[1]: *** [Makefile.build:647: theories/Init/Datatypes.vo] Error 1
```
(Full build log: https://kojipkgs.fedoraproject.org//work/tasks/7185/21097185/build.log)

```
$ nm plugins/syntax/nat_syntax_plugin.o  | grep camlLoc
                 U camlLoc__3
```

infinity0 commented:
That looks like the part of the patch which "worked by accident", and this error is not entirely unexpected.

Basically, the four places where we added `.hidden`, hides too much, and the part where we removed `name = prefix` seemed to un-hide enough things to get the ocaml testsuite passing. But possibly yet more things need to be unhidden - see the second half of Jiong Wang's comment 0018143 and the last paragraph of comment 0018150. This is something that hopefully an OCaml developer with a clearer understanding of the existing code, can propose a proper solution for.

mshinwell commented:
Yeah, I will try to pin that down exactly.

gasche commented:
Any progress on this issue?

mshinwell commented:
Sorry, I've been a bit tardy on this.  I will look at it.

xavierleroy commented:
Viewed from a distance:

The stopgap measure to fix this issue (I hope) would be to simply remove the special case for addressing symbols that are defined in the current file, whether they are morally "hidden" or not.  There will be a performance hit, but maybe it is negligible.  At any rate it looks like what the binutils people expect everyone to do, because those symbols defined in the current file might be redefined differently by dynamic loading black magic, and it's not a bug but a feature, yadda yadda.

(For the record: I do think it's a bug, just like dynamic scoping is a bug, but who am I to question the holy ELF spec and the almighty binutils developers?)

Next, we should brainstorm on how to use .hidden symbols effectively to regain some performance.  But it seems more complicated than it sounds, so let's not rush a fix.

rwmjones commented:
If I'm understanding the previous comment correctly, it would correspond to the following patch:
```
 let emit_load_symbol_addr dst s =
-  if (not !Clflags.dlcode) || Compilenv.symbol_in_current_unit s then begin
+  if (not !Clflags.dlcode) then begin
     `  adrp    {emit_reg dst}, {emit_symbol s}\n`;
```
However I tried this and it didn't work -- hanging and crashing in the test suite.

Do you have a potential simpler patch to try?

rwmjones commented:
I should say the other thing I tried was:
```
 let emit_load_symbol_addr dst s =
-  if (not !Clflags.dlcode) || Compilenv.symbol_in_current_unit s then begin
-    `  adrp    {emit_reg dst}, {emit_symbol s}\n`;
-    `  add     {emit_reg dst}, {emit_reg dst}, #:lo12:{emit_symbol s}\n`
-  end else begin
-    `  adrp    {emit_reg dst}, :got:{emit_symbol s}\n`;
-    `  ldr     {emit_reg dst}, [{emit_reg dst}, #:got_lo12:{emit_symbol s}]\n`
-  end
+  `    adrp    {emit_reg dst}, :got:{emit_symbol s}\n`;
+  `    ldr     {emit_reg dst}, [{emit_reg dst}, #:got_lo12:{emit_symbol s}]\n`
 
```
which I supposed would turn every symbol into a remote reference through the GOT, but that similarly failed badly in the test suite.

xavierleroy commented:
@rwmjones : yes, that's what I had in mind.  I'm at a loss to understand why it doesn't work and what we are supposed to do!  (Damned if you do a direct reference, damned if you go through the GOT??)

fweimer commented:
glibc runs into a similar problem and solves it by using hidden aliases. The exported symbol is never marked hidden, and all internal references go to the alias, which is hidden.

There is no other way to do this in ELF because a hidden *reference* to a global symbol makes the *definition* hidden as well.

rwmjones commented:
Based on Florian's comment above, I came up with this patch:

http://oirase.annexia.org/tmp/0001-ALTERNATE-FIX-FOR-MPR-7585.patch

It's a bit of a hack (possible understatement of the year), so here are some notes in no particular order:

 * The idea is that every global symbol is shadowed with a hidden equivalent, so for example a symbol like `camlFoo__bar` also has a second hidden symbol called `camlFoo__bar_hidden`.
 * Labels are similarly shadowed with hidden versions, so `.L1234` has a hidden version `.LH1234`.
 * The method of generating these aliases is a bit primitive.  Generated code looks like:
```
    .globl camlFoo__bar
camlFoo__bar:
    .hidden camlFoo__bar_hidden
camlFoo__bar_hidden:
    <the code>
```
 * An alternate way would be to use the gas `.set` directive, which is what GCC uses to generate aliases.  However I don't know if `.set` works with non-Linux / non-ELF.
 * Wherever `adrp` is used (except through the GOT), I use the hidden symbol or label.  Everywhere else I use the normal version.
 * Coq compiles with this patch.
 * The OCaml test suite fails in the `arith.cmm` test which is doing some weird stuff with intermediate lambda trees.  I didn't look very closely.
 * `-for-pack` seems like it might be a bit broken, and there are still linker warnings.
 * I didn't try compiling anything else apart from Coq and OCaml itself, so there could be massive breakage elsewhere.  In fact, I'd be surprised if this *isn't* the case.

rwmjones commented:
Here is a cleaned up version of the previous patch.

http://oirase.annexia.org/tmp/0001-arm64-Emit-hidden-alias-for-each-global-symbol.patch

 * Fixes -pack tests.
 * Doesn't fix the `arith.cmm` test, but that's now the only test which fails.
 * This version uses `.set` which is a slightly less ugly way to do aliases.  It may be Linux-specific though, I didn't test any other platform.
 * The hidden symbols are now called `.L1234$hidden` or `camlFoo__bar$hidden` which prevents ordinary OCaml symbols from ever being mistaken for the hidden symbols (in the previous patch you could have had a genuine symbol called `Foo.bar_hidden` which would have confused things).

rwmjones commented:
I have added this patch (https://pagure.io/fedora-ocaml/c/e7fdaf008e047c445fd7a6acf9362d8b5940bf4b?branch=fedora-27-4.05.0) to Fedora, and rebuilt many packages using it.  It will get a bit wider testing now.

infinity0 commented:
Thanks for doing this! Did you disable the arith.cmm test, or fix it, or are these Fedora builds ignoring test failures by default?

rwmjones commented:
The `arith.cmm` test is still broken with this patch and yes I had to disable the tests (just on aarch64) in the Fedora build for this reason.  The Fedora package commit is: https://src.fedoraproject.org/rpms/ocaml/c/5ea3c9844957b0013e2d74ac6ac17c714381e081?branch=f27

xavierleroy commented:
Don't obsess about the `arith.cmm` test, it's using the ocamlopt back-end in nonstandard ways, with a high probability that the test is wrong and not the code in the context of this PR.

xavierleroy commented:
I'd like to put a simple fix in 4.06 before we embark in creative uses of hidden symbols and aliases.  So, let me ask (one more time) @rwmjones and others to propose and validate a patch where the optimization for referencing locally-defined symbols is gone and all accesses go through the GOT.

rwmjones commented:
In answer to Xavier's last question, I could not come up with a working patch that made all symbols go through the GOT.  It's not as simple as just changing `emit_addressing`.

If you have a patch to make everything go through the GOT then I'll happily test it for you.

BTW if you want a quicker response, it's best to send me an email because I cannot get github to notify me by email about responses on this bug.

mshinwell commented:
I am going to see if I can construct such a patch.

mshinwell commented:
I think I have a working patch which I shall present tomorrow.

mshinwell commented:
Please see https://github.com/ocaml/ocaml/pull/1330

xavierleroy commented:
The basic workaround ( #1330 ) is merged and will go in 4.06.  We can now resume dreaming about hidden symbols and the like :-)

gasche commented:
No discussion happened since 2017. My impression is that people were interested in fixing the bug, not so much in minute details of performance of native code on arm64 backends. So I'm closing this. If people do want to work on the .hidden optimizations, they should feel free to reopen, but for now we don't have any candidates.
